diff -ruN '--exclude=.git' '--exclude=build' '--exclude=.gitignore' '--exclude=.mailmap' '--exclude=*.o' '--exclude=*.a' '--exclude=*.pyc' '--exclude=aml-user-key.sig' '--exclude=aml-user-key-v3.sig' '--exclude=aml-key' a/drivers/vpu/aml_vpu.c b/drivers/vpu/aml_vpu.c
--- a/drivers/vpu/aml_vpu.c	1970-01-01 08:00:00.000000000 +0800
+++ b/drivers/vpu/aml_vpu.c	2026-02-09 11:25:48.633084759 +0800
@@ -0,0 +1,1075 @@
+/* SPDX-License-Identifier: (GPL-2.0+ OR MIT) */
+/*
+ * drivers/vpu/aml_vpu.c
+ *
+ * Copyright (C) 2020 Amlogic, Inc. All rights reserved.
+ *
+ */
+
+#include <common.h>
+#include <asm/cpu_id.h>
+#include <config.h>
+#include <linux/kernel.h>
+#ifdef CONFIG_OF_LIBFDT
+#include <libfdt.h>
+#endif
+#include <vpu.h>
+#include "aml_vpu_reg.h"
+#include "aml_vpu.h"
+#include "aml_vpu_ctrl.h"
+
+/* v04: add txlx support */
+/* v05: add axg support */
+/* v06: add g12a support */
+/* v20180925: add tl1 support */
+/* v20190313: add sm1 support */
+/* v20200817: add t5 support */
+/* v20201022: add t5d support */
+/* v20210819: add t5w support */
+#define VPU_VERION	"v20210819"
+
+static int dts_ready = 0;
+
+struct vpu_conf_s vpu_conf = {
+	.data = NULL,
+
+	.clk_level = 0,
+	.fclk_freq = FCLK_2000M,
+};
+
+static struct vpu_data_s vpu_data_gxb = {
+	.chip_type = VPU_CHIP_GXBB,
+	.chip_name = "gxbb",
+	.clk_level_dft = CLK_LEVEL_DFT_GXBB,
+	.clk_level_max = CLK_LEVEL_MAX_GXBB,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_gxb,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_gxb,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_gx,
+
+	.module_init_table_cnt = 0,
+	.module_init_table = NULL,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_gxtvbb = {
+	.chip_type = VPU_CHIP_GXTVBB,
+	.chip_name = "gxtvbb",
+	.clk_level_dft = CLK_LEVEL_DFT_GXTVBB,
+	.clk_level_max = CLK_LEVEL_MAX_GXTVBB,
+	.gp_pll_valid = 1,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_gxb,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_gxb,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_gx,
+
+	.module_init_table_cnt = 0,
+	.module_init_table = NULL,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_gxl = {
+	.chip_type = VPU_CHIP_GXL,
+	.chip_name = "gxl",
+	.clk_level_dft = CLK_LEVEL_DFT_GXL,
+	.clk_level_max = CLK_LEVEL_MAX_GXL,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_gxb,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_gxb,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_gx,
+
+	.module_init_table_cnt = 0,
+	.module_init_table = NULL,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_gxm = {
+	.chip_type = VPU_CHIP_GXM,
+	.chip_name = "gxm",
+	.clk_level_dft = CLK_LEVEL_DFT_GXM,
+	.clk_level_max = CLK_LEVEL_MAX_GXM,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_gxb,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_gxb,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_gx,
+
+	.module_init_table_cnt = sizeof(vpu_module_init_gxm) / sizeof(struct vpu_ctrl_s),
+	.module_init_table = vpu_module_init_gxm,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_txl = {
+	.chip_type = VPU_CHIP_TXL,
+	.chip_name = "txl",
+	.clk_level_dft = CLK_LEVEL_DFT_TXLX,
+	.clk_level_max = CLK_LEVEL_MAX_TXLX,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_gxb,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_gxb,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_gx,
+
+	.module_init_table_cnt = 0,
+	.module_init_table = NULL,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_txlx = {
+	.chip_type = VPU_CHIP_TXLX,
+	.chip_name = "txlx",
+	.clk_level_dft = CLK_LEVEL_DFT_TXLX,
+	.clk_level_max = CLK_LEVEL_MAX_TXLX,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_gxb,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_gxb,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_gx,
+
+	.module_init_table_cnt = sizeof(vpu_module_init_txlx) / sizeof(struct vpu_ctrl_s),
+	.module_init_table = vpu_module_init_txlx,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_axg = {
+	.chip_type = VPU_CHIP_AXG,
+	.chip_name = "axg",
+	.clk_level_dft = CLK_LEVEL_DFT_AXG,
+	.clk_level_max = CLK_LEVEL_MAX_AXG,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_gxb,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_axg,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_gx,
+
+	.module_init_table_cnt = 0,
+	.module_init_table = NULL,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_txhd = {
+	.chip_type = VPU_CHIP_TXHD,
+	.chip_name = "txhd",
+	.clk_level_dft = CLK_LEVEL_DFT_TXHD,
+	.clk_level_max = CLK_LEVEL_MAX_TXHD,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_gxb,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_txhd,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_txhd,
+
+	.module_init_table_cnt = 0,
+	.module_init_table = NULL,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_g12a = {
+	.chip_type = VPU_CHIP_G12A,
+	.chip_name = "g12a",
+	.clk_level_dft = CLK_LEVEL_DFT_G12A,
+	.clk_level_max = CLK_LEVEL_MAX_G12A,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_g12a,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_g12a,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_gx,
+
+	.module_init_table_cnt = sizeof(vpu_module_init_g12a) / sizeof(struct vpu_ctrl_s),
+	.module_init_table = vpu_module_init_g12a,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_g12b = {
+	.chip_type = VPU_CHIP_G12B,
+	.chip_name = "g12b",
+	.clk_level_dft = CLK_LEVEL_DFT_G12A,
+	.clk_level_max = CLK_LEVEL_MAX_G12A,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_g12a,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_g12b,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_gx,
+
+	.module_init_table_cnt = sizeof(vpu_module_init_g12a) / sizeof(struct vpu_ctrl_s),
+	.module_init_table = vpu_module_init_g12a,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_tl1 = {
+	.chip_type = VPU_CHIP_TL1,
+	.chip_name = "tl1",
+	.clk_level_dft = CLK_LEVEL_DFT_G12A,
+	.clk_level_max = CLK_LEVEL_MAX_G12A,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_g12a,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_tl1,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_gxb,
+	.reset_table = vpu_reset_tl1,
+
+	.module_init_table_cnt = 0,
+	.module_init_table = NULL,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_sm1 = {
+	.chip_type = VPU_CHIP_SM1,
+	.chip_name = "sm1",
+	.clk_level_dft = CLK_LEVEL_DFT_G12A,
+	.clk_level_max = CLK_LEVEL_MAX_G12A,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_g12a,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_tl1,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_sm1,
+	.reset_table = vpu_reset_gx,
+
+	.module_init_table_cnt = sizeof(vpu_module_init_g12a) / sizeof(struct vpu_ctrl_s),
+	.module_init_table = vpu_module_init_g12a,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_tm2 = {
+	.chip_type = VPU_CHIP_TM2,
+	.chip_name = "tm2",
+	.clk_level_dft = CLK_LEVEL_DFT_G12A,
+	.clk_level_max = CLK_LEVEL_MAX_G12A,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = VPU_PWR_ID_INVALID,
+
+	.fclk_div_table = fclk_div_table_g12a,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_tm2,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_sm1,
+	.reset_table = vpu_reset_tl1,
+
+	.module_init_table_cnt = sizeof(vpu_module_init_tm2) / sizeof(struct vpu_ctrl_s),
+	.module_init_table = vpu_module_init_tm2,
+
+	.power_on = vpu_power_on,
+	.power_off = vpu_power_off,
+};
+
+static struct vpu_data_s vpu_data_t5 = {
+	.chip_type = VPU_CHIP_T5,
+	.chip_name = "t5",
+	.clk_level_dft = CLK_LEVEL_DFT_G12A,
+	.clk_level_max = CLK_LEVEL_MAX_G12A,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = PM_VPU_HDMI,
+
+	.fclk_div_table = fclk_div_table_g12a,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = vpu_mem_pd_tm2,
+	.power_table = vpu_pwr_gxb,
+	.iso_table = vpu_iso_sm1,
+	.reset_table = vpu_reset_tl1,
+
+	.module_init_table_cnt = sizeof(vpu_module_init_tm2) / sizeof(struct vpu_ctrl_s),
+	.module_init_table = vpu_module_init_tm2,
+
+	.power_on = vpu_power_on_new,
+	.power_off = vpu_power_off_new,
+};
+
+static struct vpu_data_s vpu_data_t5d = {
+	.chip_type = VPU_CHIP_T5D,
+	.chip_name = "t5d",
+	.clk_level_dft = CLK_LEVEL_DFT_T5D,
+	.clk_level_max = CLK_LEVEL_MAX_T5D,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CNTL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = PM_VPU_HDMI,
+
+	.fclk_div_table = fclk_div_table_g12a,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = NULL,
+	.power_table = NULL,
+	.iso_table = NULL,
+	.reset_table = NULL,
+
+	.module_init_table_cnt = sizeof(vpu_module_init_tm2) / sizeof(struct vpu_ctrl_s),
+	.module_init_table = vpu_module_init_tm2,
+
+	.power_on = vpu_power_on_new,
+	.power_off = vpu_power_off_new,
+};
+
+static struct vpu_data_s vpu_data_t5w = {
+	.chip_type = VPU_CHIP_T5W,
+	.chip_name = "t5w",
+	.clk_level_dft = CLK_LEVEL_DFT_G12A,
+	.clk_level_max = CLK_LEVEL_MAX_G12A,
+	.gp_pll_valid = 0,
+
+	.vpu_clk_reg = HHI_VPU_CLK_CTRL,
+	.vapb_clk_reg = HHI_VAPBCLK_CNTL,
+	.vid_clk_reg = HHI_VID_CLK_CNTL2,
+
+	.pwrctrl_id = PM_VPU_HDMI,
+
+	.fclk_div_table = fclk_div_table_g12a,
+	.vpu_clk_table = vpu_clk_table,
+
+	.mem_pd_table = NULL,
+	.power_table = NULL,
+	.iso_table = NULL,
+	.reset_table = NULL,
+
+	.module_init_table_cnt = sizeof(vpu_module_init_tm2) / sizeof(struct vpu_ctrl_s),
+	.module_init_table = vpu_module_init_tm2,
+
+	.power_on = NULL,
+	.power_off = NULL,
+//	.power_on = vpu_power_on_new,
+//	.power_off = vpu_power_off_new,
+};
+
+static void vpu_chip_detect(void)
+{
+	unsigned int cpu_type;
+
+	cpu_type = get_cpu_id().family_id;
+	switch (cpu_type) {
+	case MESON_CPU_MAJOR_ID_GXBB:
+		vpu_conf.data = &vpu_data_gxb;
+		break;
+	case MESON_CPU_MAJOR_ID_GXTVBB:
+		vpu_conf.data = &vpu_data_gxtvbb;
+		break;
+	case MESON_CPU_MAJOR_ID_GXL:
+		vpu_conf.data = &vpu_data_gxl;
+		break;
+	case MESON_CPU_MAJOR_ID_GXM:
+		vpu_conf.data = &vpu_data_gxm;
+		break;
+	case MESON_CPU_MAJOR_ID_TXL:
+		vpu_conf.data = &vpu_data_txl;
+		break;
+	case MESON_CPU_MAJOR_ID_TXLX:
+		vpu_conf.data = &vpu_data_txlx;
+		break;
+	case MESON_CPU_MAJOR_ID_AXG:
+		vpu_conf.data = &vpu_data_axg;
+		break;
+	case MESON_CPU_MAJOR_ID_TXHD:
+		vpu_conf.data = &vpu_data_txhd;
+		break;
+	case MESON_CPU_MAJOR_ID_G12A:
+		vpu_conf.data = &vpu_data_g12a;
+		break;
+	case MESON_CPU_MAJOR_ID_G12B:
+		vpu_conf.data = &vpu_data_g12b;
+		break;
+	case MESON_CPU_MAJOR_ID_TL1:
+		vpu_conf.data = &vpu_data_tl1;
+		break;
+	case MESON_CPU_MAJOR_ID_SM1:
+		vpu_conf.data = &vpu_data_sm1;
+		break;
+	case MESON_CPU_MAJOR_ID_TM2:
+		vpu_conf.data = &vpu_data_tm2;
+		break;
+	case MESON_CPU_MAJOR_ID_T5:
+		vpu_conf.data = &vpu_data_t5;
+		break;
+	case MESON_CPU_MAJOR_ID_T5D:
+		vpu_conf.data = &vpu_data_t5d;
+		break;
+	case MESON_CPU_MAJOR_ID_T5W:
+		vpu_conf.data = &vpu_data_t5w;
+		break;
+	default:
+		vpu_conf.data = &vpu_data_t5w;
+		break;
+	}
+
+	strcpy(vpu_conf.drv_version, VPU_VERION);
+
+	if (vpu_conf.data == NULL) {
+		VPUERR("invalid vpu\n");
+		return ;
+	} else {
+#ifdef CONFIG_VPU_CLK_LEVEL_DFT
+		vpu_conf.data->clk_level_dft = CONFIG_VPU_CLK_LEVEL_DFT;
+#endif
+
+#ifdef VPU_DEBUG_PRINT
+		VPUPR("driver version: %s\n", vpu_conf.drv_version);
+		VPUPR("detect chip type: %d\n", vpu_conf.data->chip_type);
+		VPUPR("clk_level default: %d(%dHz), max: %d(%dHz)\n",
+			vpu_conf.data->clk_level_dft,
+			(vpu_conf.data->vpu_clk_table + vpu_conf.data->clk_level_dft)->freq,
+			(vpu_conf.data->clk_level_max - 1),
+			(vpu_conf.data->vpu_clk_table + (vpu_conf.data->clk_level_max - 1))->freq);
+#endif
+	}
+}
+
+static int vpu_check(void)
+{
+	int ret = -1;
+
+	if (vpu_conf.data == NULL) {
+		VPUERR("invalid vpu\n");
+		return -1;
+	}
+
+	switch (vpu_conf.data->chip_type) {
+	case VPU_CHIP_GXBB:
+	case VPU_CHIP_GXTVBB:
+	case VPU_CHIP_GXL:
+	case VPU_CHIP_GXM:
+	case VPU_CHIP_TXL:
+	case VPU_CHIP_TXLX:
+	case VPU_CHIP_AXG:
+	case VPU_CHIP_TXHD:
+	case VPU_CHIP_G12A:
+	case VPU_CHIP_G12B:
+	case VPU_CHIP_TL1:
+	case VPU_CHIP_SM1:
+	case VPU_CHIP_TM2:
+	case VPU_CHIP_T5:
+	case VPU_CHIP_T5D:
+	case VPU_CHIP_T5W:
+		ret = 0;
+		break;
+	default:
+		VPUERR("invalid vpu for current chip\n");
+		break;
+	}
+
+	return ret;
+}
+
+static unsigned int get_vpu_clk_level(unsigned int video_clk)
+{
+	unsigned clk_level;
+	int i;
+
+	for (i = 0; i < vpu_conf.data->clk_level_max; i++) {
+		if (video_clk <= (vpu_conf.data->vpu_clk_table + i)->freq)
+			break;
+	}
+	clk_level = i;
+
+	return clk_level;
+}
+
+static unsigned long get_fclk_div_freq(unsigned int mux_id)
+{
+	struct fclk_div_s *fclk_div;
+	unsigned long fclk, div, clk_source = 0;
+	unsigned int i;
+
+	fclk = vpu_conf.fclk_freq * 1000000;
+
+	for (i = 0; i < FCLK_DIV_MAX; i++) {
+		fclk_div = vpu_conf.data->fclk_div_table + i;
+		if (fclk_div->fclk_id == mux_id) {
+			div = fclk_div->fclk_div;
+			clk_source = ((fclk * 100 / div) + 99) / 100;
+			break;
+		}
+		if (fclk_div->fclk_id == FCLK_DIV_MAX)
+			break;
+	}
+
+	return clk_source;
+}
+
+static unsigned int get_vpu_clk_mux_id(void)
+{
+	struct fclk_div_s *fclk_div;
+	unsigned int i, mux, mux_id;
+
+	mux = vpu_hiu_getb(vpu_conf.data->vpu_clk_reg, 9, 3);
+	mux_id = mux;
+	for (i = 0; i < FCLK_DIV_MAX; i++) {
+		fclk_div = vpu_conf.data->fclk_div_table + i;
+		if (fclk_div->fclk_mux == mux) {
+			mux_id = fclk_div->fclk_id;
+			break;
+		}
+		if (fclk_div->fclk_id == FCLK_DIV_MAX)
+			break;
+	}
+
+	return mux_id;
+}
+
+static unsigned int get_vpu_clk_mux(unsigned int mux_id)
+{
+	struct fclk_div_s *fclk_div;
+	unsigned int i, mux;
+
+	mux = mux_id;
+	for (i = 0; i < FCLK_DIV_MAX; i++) {
+		fclk_div = vpu_conf.data->fclk_div_table + i;
+		if (fclk_div->fclk_id == mux_id) {
+			mux = fclk_div->fclk_mux;
+			break;
+		}
+		if (fclk_div->fclk_id == FCLK_DIV_MAX)
+			break;
+	}
+
+	return mux;
+}
+
+static unsigned int get_vpu_clk(void)
+{
+	unsigned long clk_freq;
+	unsigned long clk_source, div;
+	unsigned int mux_id;
+
+	mux_id = get_vpu_clk_mux_id();
+	switch (mux_id) {
+	case FCLK_DIV4:
+	case FCLK_DIV3:
+	case FCLK_DIV5:
+	case FCLK_DIV7:
+		clk_source = get_fclk_div_freq(mux_id);
+		break;
+	case GPLL_CLK:
+		clk_source = (vpu_conf.data->vpu_clk_table + 8)->freq;
+		break;
+	default:
+		clk_source = 0;
+		break;
+	}
+
+	div = vpu_hiu_getb(vpu_conf.data->vpu_clk_reg, 0, 7) + 1;
+	clk_freq = ((clk_source * 100 / div) + 99) / 100;
+
+	return (unsigned int)clk_freq;
+}
+
+static int switch_gp1_pll_gxtvbb(int flag)
+{
+	int cnt = 100;
+	int ret = 0;
+
+	if (flag) { /* enable gp1_pll */
+		/* GP1 DPLL 696MHz output*/
+		vpu_hiu_write(HHI_GP1_PLL_CNTL2, 0x69c80000);
+		vpu_hiu_write(HHI_GP1_PLL_CNTL3, 0x0a5590c4);
+		vpu_hiu_write(HHI_GP1_PLL_CNTL4, 0x0000500d);
+		vpu_hiu_write(HHI_GP1_PLL_CNTL, 0x6a01023a);
+		vpu_hiu_write(HHI_GP1_PLL_CNTL, 0x4a01023a);
+		do {
+			udelay(10);
+			vpu_hiu_setb(HHI_GP1_PLL_CNTL, 1, 29, 1); /* reset */
+			udelay(50);
+			/* release reset */
+			vpu_hiu_setb(HHI_GP1_PLL_CNTL, 0, 29, 1);
+			udelay(50);
+			cnt--;
+			if (cnt == 0)
+				break;
+		} while (vpu_hiu_getb(HHI_GP1_PLL_CNTL, 31, 1) == 0);
+		if (vpu_hiu_getb(HHI_GP1_PLL_CNTL, 31, 1) == 0) {
+			ret = -1;
+			vpu_hiu_setb(HHI_GP1_PLL_CNTL, 0, 30, 1);
+			VPUERR("GP1_PLL lock failed\n");
+		}
+	} else { /* disable gp1_pll */
+		vpu_hiu_setb(HHI_GP1_PLL_CNTL, 0, 30, 1);
+	}
+
+	return ret;
+}
+
+static int switch_gp_pll(int flag)
+{
+	int ret = -1;
+
+	if (vpu_conf.data->gp_pll_valid == 0)
+		return ret;
+
+	switch (vpu_conf.data->chip_type) {
+	case VPU_CHIP_GXTVBB:
+		ret = switch_gp1_pll_gxtvbb(flag);
+		break;
+	default:
+		break;
+	}
+
+	return ret;
+}
+
+/* unit: MHz */
+#define VPU_CLKB_MAX    350
+static int adjust_vpu_clk(unsigned int clk_level)
+{
+	struct vpu_clk_s *clk_table;
+	unsigned int vpu_clk;
+	unsigned int mux, div, clk_reg, vapb_reg;
+	int ret = 0;
+
+	/* vpu clk */
+	clk_table = vpu_conf.data->vpu_clk_table + clk_level;
+	mux = get_vpu_clk_mux(clk_table->mux);
+	if (mux == GPLL_CLK) {
+		ret = switch_gp_pll(1);
+		if (ret) {
+			clk_level = vpu_conf.data->clk_level_dft;
+			VPUERR("adjust_vpu_clk: gp pll error, use default clk\n");
+		}
+	}
+	vpu_conf.clk_level = clk_level;
+	clk_table = vpu_conf.data->vpu_clk_table + clk_level;
+
+	vpu_clk = clk_table->freq;
+	mux = get_vpu_clk_mux(clk_table->mux);
+	div = clk_table->div;
+	clk_reg = vpu_conf.data->vpu_clk_reg;
+	vapb_reg = vpu_conf.data->vapb_clk_reg;
+
+	vpu_hiu_write(clk_reg, ((mux << 9) | (div << 0)));
+	vpu_hiu_setb(clk_reg, 1, 8, 1);
+
+	/* vpu clkb if needed */
+	switch (vpu_conf.data->chip_type) {
+	case VPU_CHIP_GXBB:
+	case VPU_CHIP_GXTVBB:
+	case VPU_CHIP_GXL:
+	case VPU_CHIP_GXM:
+	case VPU_CHIP_TXL:
+		if (vpu_clk >= (VPU_CLKB_MAX * 1000000))
+			vpu_hiu_write(HHI_VPU_CLKB_CNTL, ((1 << 8) | (1 << 0)));
+		else
+			vpu_hiu_write(HHI_VPU_CLKB_CNTL, ((1 << 8) | (0 << 0)));
+		break;
+	default:
+		break;
+	}
+
+	/* vapb clk */
+	if (vpu_clk >= 250000000) {
+		vpu_hiu_write(vapb_reg, (1 << 30) | /* turn on ge2d clock */
+					(0 << 9)  | /* clk_sel    //250Mhz */
+					(1 << 0));  /* clk_div */
+	} else {
+		vpu_hiu_write(vapb_reg, (1 << 30) | /* turn on ge2d clock */
+					(clk_table->mux << 9)  | /* clk_sel */
+					(div << 0));  /* clk_div */
+	}
+	vpu_hiu_setb(vapb_reg, 1, 8, 1);
+
+	VPUPR("set clk: %uHz, readback: %uHz(0x%x)\n",
+		vpu_clk, get_vpu_clk(), (vpu_hiu_read(clk_reg)));
+	return ret;
+}
+
+static int set_vpu_clk(unsigned int vclk)
+{
+	int ret = 0;
+	unsigned int clk_level;
+
+#ifdef VPU_DEBUG_PRINT
+	VPUPR("%s\n", __func__);
+#endif
+
+	if (vclk >= 100) /* regard as vpu_clk */
+		clk_level = get_vpu_clk_level(vclk);
+	else /* regard as clk_level */
+		clk_level = vclk;
+
+	if (clk_level >= vpu_conf.data->clk_level_max) {
+		clk_level = vpu_conf.data->clk_level_dft;
+		VPUPR("clk out of supported range, set to default\n");
+	}
+
+	ret = adjust_vpu_clk(clk_level);
+
+#ifdef VPU_DEBUG_PRINT
+	VPUPR("%s finish\n", __func__);
+#endif
+
+	return ret;
+}
+
+static int get_vpu_config(char *dt_addr)
+{
+#ifdef CONFIG_OF_LIBFDT
+	int nodeoffset;
+	char *propdata;
+#endif
+
+	if (dts_ready == 1) {
+#ifdef CONFIG_OF_LIBFDT
+		nodeoffset = fdt_path_offset(dt_addr, "/vpu");
+		if (nodeoffset < 0) {
+			VPUERR("not find /vpu node in dts %s\n", fdt_strerror(nodeoffset));
+			return -1;
+		}
+
+		propdata = (char *)fdt_getprop(dt_addr, nodeoffset, "clk_level", NULL);
+		if (propdata == NULL) {
+			vpu_conf.clk_level = vpu_conf.data->clk_level_dft;
+			VPUERR("don't find clk_level in dts, set to default\n");
+		} else {
+			vpu_conf.clk_level = (unsigned short)(be32_to_cpup((u32*)propdata));
+			if (vpu_conf.clk_level >= vpu_conf.data->clk_level_max) {
+				VPUERR("clk_level in dts is out of support, set to default\n");
+				vpu_conf.clk_level = vpu_conf.data->clk_level_dft;
+			}
+		}
+		VPUPR("clk_level in dts: %u\n", vpu_conf.clk_level);
+#endif
+	} else {
+		vpu_conf.clk_level = vpu_conf.data->clk_level_dft;
+		VPUPR("clk_level = %u\n", vpu_conf.clk_level);
+	}
+
+	return 0;
+}
+
+int vpu_probe(void)
+{
+	unsigned long dtb_mem;
+	char *dt_addr;
+	int ret;
+
+	dts_ready = 0;
+
+	dtb_mem = getenv_ulong("dtb_mem_addr", 16, 0);
+	if (dtb_mem) {
+		dt_addr = (char *)dtb_mem;
+	} else {
+		VPUPR("no dtb_mem_addr\n");
+#ifdef CONFIG_DTB_MEM_ADDR
+		dt_addr = (char *)CONFIG_DTB_MEM_ADDR;
+#endif
+	}
+#ifdef CONFIG_OF_LIBFDT
+	if (dt_addr) {
+		ret = fdt_check_header((const void *)dt_addr);
+		if (ret < 0) {
+			VPUERR("vpu: check dts: %s, load default parameters\n",
+				fdt_strerror(ret));
+		} else {
+			dts_ready = 1;
+		}
+	}
+#endif
+
+	vpu_chip_detect();
+	if (vpu_check())
+		return -1;
+
+	ret = get_vpu_config(dt_addr);
+	if (vpu_conf.data->power_on)
+		vpu_conf.data->power_on();
+	set_vpu_clk(vpu_conf.clk_level);
+
+	/* vpu module init off, for power save, and special module init */
+	vpu_mem_pd_init_off();
+	vpu_module_init_config();
+
+	return ret;
+}
+
+int vpu_remove(void)
+{
+	if (vpu_check())
+		return -1;
+
+	VPUPR("vpu remove\n");
+	if (vpu_conf.data->power_off)
+		vpu_conf.data->power_off();
+	return 0;
+}
+
+static void vpu_clk_switch(void)
+{
+	struct vpu_clk_s *clk_table;
+	unsigned int mux, div, clk_reg;
+
+	clk_reg = vpu_conf.data->vpu_clk_reg;
+
+	/* step 1: switch to 2nd vpu clk patch */
+	clk_table = vpu_conf.data->vpu_clk_table + vpu_conf.data->clk_level_dft;
+	mux = get_vpu_clk_mux(clk_table->mux);
+	vpu_hiu_setb(clk_reg, mux, 25, 3);
+	div = clk_table->div;
+	vpu_hiu_setb(clk_reg, div, 16, 7);
+	vpu_hiu_setb(clk_reg, 1, 24, 1);
+	vpu_hiu_setb(clk_reg, 1, 31, 1);
+	udelay(10);
+	/* step 2: adjust 1st vpu clk frequency */
+	clk_table = vpu_conf.data->vpu_clk_table + vpu_conf.clk_level;
+	vpu_hiu_setb(clk_reg, 0, 8, 1);
+	mux = get_vpu_clk_mux(clk_table->mux);
+	vpu_hiu_setb(clk_reg, mux, 9, 3);
+	div = clk_table->div;
+	vpu_hiu_setb(clk_reg, div, 0, 7);
+	vpu_hiu_setb(clk_reg, 1, 8, 1);
+	udelay(20);
+	/* step 3: switch back to 1st vpu clk patch */
+	vpu_hiu_setb(clk_reg, 0, 31, 1);
+	vpu_hiu_setb(clk_reg, 0, 24, 1);
+}
+
+int vpu_clk_change(int level)
+{
+	struct vpu_clk_s *clk_table;
+	unsigned int vpu_clk;
+
+	if (vpu_check())
+		return -1;
+
+	if (level >= 100) /* regard as vpu_clk */
+		level = get_vpu_clk_level(level);
+
+	if (level >= vpu_conf.data->clk_level_max) {
+		clk_table = vpu_conf.data->vpu_clk_table + (vpu_conf.data->clk_level_max - 1);
+		VPUPR("clk out of supported range\n");
+		VPUPR("clk max level: %u(%uHz)\n",
+			(vpu_conf.data->clk_level_max - 1),
+			clk_table->freq);
+		return -1;
+	}
+
+	vpu_conf.clk_level = level;
+	clk_table = vpu_conf.data->vpu_clk_table + vpu_conf.clk_level;
+	vpu_clk = clk_table->freq;
+
+	vpu_clk_switch();
+
+	VPUPR("set clk: %uHz, readback: %uHz(0x%x)\n",
+		vpu_clk, get_vpu_clk(), vpu_hiu_read(vpu_conf.data->vpu_clk_reg));
+	return 0;
+}
+
+void vpu_clk_get(void)
+{
+	if (vpu_check())
+		return;
+
+	VPUPR("clk_level: %u, clk: %uHz, reg: 0x%x\n",
+		vpu_conf.clk_level, get_vpu_clk(),
+		vpu_hiu_read(vpu_conf.data->vpu_clk_reg));
+}
+
+void vpu_info_print(void)
+{
+	if (vpu_check())
+		return;
+
+	printf("vpu driver version: %s\n", vpu_conf.drv_version);
+	printf("detect chip type: %d(%s)\n",
+		vpu_conf.data->chip_type,
+		vpu_conf.data->chip_name);
+	printf("clk_level:         %d(%dHz)\n"
+		"clk_level default: %d(%dHz)\n"
+		"clk_level max:     %d(%dHz)\n",
+		vpu_conf.clk_level,
+		(vpu_conf.data->vpu_clk_table + vpu_conf.clk_level)->freq,
+		vpu_conf.data->clk_level_dft,
+		(vpu_conf.data->vpu_clk_table + vpu_conf.data->clk_level_dft)->freq,
+		(vpu_conf.data->clk_level_max - 1),
+		(vpu_conf.data->vpu_clk_table + (vpu_conf.data->clk_level_max - 1))->freq);
+}
+
+static unsigned int vcbus_reg[] = {
+	VENC_VDAC_TST_VAL, /* 0x1b7f */
+	ENCP_DVI_HSO_BEGIN, /* 0x1c30 */
+	VPP_DUMMY_DATA, /* 0x1d00 */
+	VPU_VPU_PWM_V0, /* 0x2730 */
+};
+
+void vcbus_test(void)
+{
+	unsigned int val;
+	unsigned int temp;
+	int i,j;
+
+	if (vpu_check())
+		return;
+
+	VPUPR("vcbus test:\n");
+	for (i = 0; i < ARRAY_SIZE(vcbus_reg); i++) {
+		for (j = 0; j < 24; j++) {
+			val = vpu_vcbus_read(vcbus_reg[i]);
+			printf("%02d read 0x%04x=0x%08x\n", j, vcbus_reg[i], val);
+		}
+		printf("\n");
+	}
+	temp = 0x5a5a5a5a;
+	for (i = 0; i < ARRAY_SIZE(vcbus_reg); i++) {
+		vpu_vcbus_write(vcbus_reg[i], temp);
+		val = vpu_vcbus_read(vcbus_reg[i]);
+		printf("write 0x%04x=0x%08x, readback: 0x%08x\n", vcbus_reg[i], temp, val);
+	}
+	for (i = 0; i < ARRAY_SIZE(vcbus_reg); i++) {
+		for (j = 0; j < 24; j++) {
+			val = vpu_vcbus_read(vcbus_reg[i]);
+			printf("%02d read 0x%04x=0x%08x\n", j, vcbus_reg[i], val);
+		}
+		printf("\n");
+	}
+}
diff -ruN '--exclude=.git' '--exclude=build' '--exclude=.gitignore' '--exclude=.mailmap' '--exclude=*.o' '--exclude=*.a' '--exclude=*.pyc' '--exclude=aml-user-key.sig' '--exclude=aml-user-key-v3.sig' '--exclude=aml-key' a/drivers/vpu/aml_vpu_ctrl.h b/drivers/vpu/aml_vpu_ctrl.h
--- a/drivers/vpu/aml_vpu_ctrl.h	1970-01-01 08:00:00.000000000 +0800
+++ b/drivers/vpu/aml_vpu_ctrl.h	2026-02-09 11:25:48.633084759 +0800
@@ -0,0 +1,277 @@
+/* SPDX-License-Identifier: (GPL-2.0+ OR MIT) */
+/*
+ * drivers/vpu/aml_vpu_ctrl.h
+ *
+ * Copyright (C) 2020 Amlogic, Inc. All rights reserved.
+ *
+ */
+
+#ifndef __VPU_CTRL_H__
+#define __VPU_CTRL_H__
+#include "aml_vpu.h"
+
+/* #define LIMIT_VPU_CLK_LOW */
+
+/* ************************************************ */
+/* VPU frequency table, important. DO NOT modify!! */
+/* ************************************************ */
+/* fixed pll frequency */
+#define FCLK_2000M    2000 /* unit: MHz */
+
+/* GXBB */
+/* freq max=666M, default=666M */
+#define CLK_LEVEL_DFT_GXBB     7
+#define CLK_LEVEL_MAX_GXBB     8
+/* GXTVBB */
+/* freq max=666M, default=666M */
+#define CLK_LEVEL_DFT_GXTVBB   7
+#define CLK_LEVEL_MAX_GXTVBB   8
+/* GXL */
+/* freq max=666M, default=666M */
+#define CLK_LEVEL_DFT_GXL      7
+#define CLK_LEVEL_MAX_GXL      8
+/* GXM */
+/* freq max=666M, default=666M */
+#define CLK_LEVEL_DFT_GXM      7
+#define CLK_LEVEL_MAX_GXM      8
+/* TXL */
+/* freq max=666M, default=666M */
+#define CLK_LEVEL_DFT_TXL      7
+#define CLK_LEVEL_MAX_TXL      8
+
+/* TXLX */
+/* freq max=666M, default=666M */
+#define CLK_LEVEL_DFT_TXLX     7
+#define CLK_LEVEL_MAX_TXLX     8
+
+/* AXG */
+/* freq max=250M, default=250M */
+#define CLK_LEVEL_DFT_AXG      3
+#define CLK_LEVEL_MAX_AXG      4
+
+/* TXHD */
+/* freq max=400M, default=400M */
+#define CLK_LEVEL_DFT_TXHD     5
+#define CLK_LEVEL_MAX_TXHD     6
+
+/* G12A */
+/* freq max=666M, default=666M */
+#define CLK_LEVEL_DFT_G12A     7
+#define CLK_LEVEL_MAX_G12A     8
+
+/* T5D */
+/* freq max=400M, default=400M */
+#define CLK_LEVEL_DFT_T5D     5
+#define CLK_LEVEL_MAX_T5D     6
+
+
+/* vpu clk setting */
+enum vpu_mux_e {
+	FCLK_DIV4 = 0,
+	FCLK_DIV3,
+	FCLK_DIV5,
+	FCLK_DIV7,
+	MPLL_CLK1,
+	VID_PLL_CLK,
+	VID2_PLL_CLK,
+	GPLL_CLK,
+	FCLK_DIV_MAX,
+};
+
+static struct fclk_div_s fclk_div_table_gxb[] = {
+	/* id,         mux,  div */
+	{FCLK_DIV4,    0,    4},
+	{FCLK_DIV3,    1,    3},
+	{FCLK_DIV5,    2,    5},
+	{FCLK_DIV7,    3,    7},
+	{FCLK_DIV_MAX, 8,    1},
+};
+
+static struct fclk_div_s fclk_div_table_g12a[] = {
+	/* id,         mux,  div */
+	{FCLK_DIV3,    0,    3},
+	{FCLK_DIV4,    1,    4},
+	{FCLK_DIV5,    2,    5},
+	{FCLK_DIV7,    3,    7},
+	{FCLK_DIV_MAX, 8,    1},
+};
+
+static struct vpu_clk_s vpu_clk_table[] = {
+	/* frequency   clk_mux       div */
+	{100000000,    FCLK_DIV5,    3}, /* 0 */
+	{166667000,    FCLK_DIV3,    3}, /* 1 */
+	{200000000,    FCLK_DIV5,    1}, /* 2 */
+	{250000000,    FCLK_DIV4,    1}, /* 3 */
+	{333333000,    FCLK_DIV3,    1}, /* 4 */
+	{400000000,    FCLK_DIV5,    0}, /* 5 */
+	{500000000,    FCLK_DIV4,    0}, /* 6 */
+	{666667000,    FCLK_DIV3,    0}, /* 7 */
+	{696000000,    GPLL_CLK,     0}, /* 8 */ /* invalid */
+	{850000000,    GPLL_CLK,     0}, /* 9 */ /* invalid */
+};
+
+/* ******************************************************* */
+/*              VPU memory power down table                */
+/* ******************************************************* */
+static struct vpu_ctrl_s vpu_mem_pd_gxb[] = {
+	/* reg,                val, bit, len */
+	{HHI_VPU_MEM_PD_REG0,  1,   0,   32},
+	{HHI_VPU_MEM_PD_REG1,  1,   0,   32},
+	{VPU_REG_END,          0,   0,    0},
+};
+
+static struct vpu_ctrl_s vpu_mem_pd_axg[] = {
+	/* reg,                val, bit, len */
+	{HHI_VPU_MEM_PD_REG0,  1,   0,   32},
+	{VPU_REG_END,          0,   0,    0},
+};
+
+static struct vpu_ctrl_s vpu_mem_pd_txhd[] = {
+	/* reg,               val, bit, len */
+	{HHI_VPU_MEM_PD_REG0, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG1, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG2, 1,   2,    2}, /* tcon memory */
+	{VPU_REG_END,         0,   0,    0},
+};
+
+static struct vpu_ctrl_s vpu_mem_pd_g12a[] = {
+	/* reg,               val, bit, len */
+	{HHI_VPU_MEM_PD_REG0, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG1, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG2, 1,   0,    2},
+	{HHI_VPU_MEM_PD_REG2, 1,   4,   14},
+	{HHI_VPU_MEM_PD_REG2, 1,  30,    2},
+	{VPU_REG_END,         0,   0,    0},
+};
+
+static struct vpu_ctrl_s vpu_mem_pd_g12b[] = {
+	/* reg,               val, bit, len */
+	{HHI_VPU_MEM_PD_REG0, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG1, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG2, 1,   0,    2},
+	{HHI_VPU_MEM_PD_REG2, 1,   4,   14},
+	{HHI_VPU_MEM_PD_REG2, 1,  22,    2},
+	{HHI_VPU_MEM_PD_REG2, 1,  30,    2},
+	{VPU_REG_END,         0,   0,    0},
+};
+
+static struct vpu_ctrl_s vpu_mem_pd_tl1[] = {
+	/* reg,               val, bit, len */
+	{HHI_VPU_MEM_PD_REG0, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG1, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG2, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG3, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG4, 1,   0,    4},
+	{VPU_REG_END,         0,   0,    0},
+};
+
+static struct vpu_ctrl_s vpu_mem_pd_tm2[] = {
+	/* reg,               val, bit, len */
+	{HHI_VPU_MEM_PD_REG0, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG1, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG2, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG3, 1,   0,   32},
+	{HHI_VPU_MEM_PD_REG4, 1,   0,    8},
+	{VPU_REG_END,         0,   0,    0},
+};
+
+/* ******************************************************* */
+/*              VPU_HDMI POWER               */
+/* ******************************************************* */
+static struct vpu_ctrl_s vpu_pwr_gxb[] = {
+	/* reg,                val, bit, len */
+	{AO_RTI_GEN_PWR_SLEEP0,  1,   8,   1},
+	{VPU_REG_END,            0,   0,   0},
+};
+
+/* ******************************************************* */
+/*              VPU_HDMI ISO                */
+/* ******************************************************* */
+static struct vpu_ctrl_s vpu_iso_gxb[] = {
+	/* reg,                val, bit, len */
+	{AO_RTI_GEN_PWR_SLEEP0,  1,   9,   1},
+	{VPU_REG_END,            0,   0,   0},
+};
+
+static struct vpu_ctrl_s vpu_iso_sm1[] = {
+	/* reg,                val, bit, len */
+	{AO_RTI_GEN_PWR_ISO0,    1,   8,   1},
+	{VPU_REG_END,            0,   0,   0},
+};
+
+/* ******************************************************* */
+/*                 VPU module init table                 */
+/* ******************************************************* */
+static struct vpu_ctrl_s vpu_module_init_gxm[] = {
+	/* reg,                     val,        bit, len */
+	{DOLBY_CORE1_CLKGATE_CTRL,  0x55555555, 0,   32},
+	{DOLBY_CORE2A_CLKGATE_CTRL, 0x55555555, 0,   32},
+	{DOLBY_CORE3_CLKGATE_CTRL,  0x55555555, 0,   32},
+	{VPU_REG_END,               0,          0,   0},
+};
+
+static struct vpu_ctrl_s vpu_module_init_txlx[] = {
+	/* reg,                     val, bit, len */
+	{DOLBY_TV_CLKGATE_CTRL,     1,   10,  2},
+	{DOLBY_TV_CLKGATE_CTRL,     1,   2,   2},
+	{DOLBY_TV_CLKGATE_CTRL,     1,   4,   2},
+	{DOLBY_CORE2A_CLKGATE_CTRL, 1,   10,  2},
+	{DOLBY_CORE2A_CLKGATE_CTRL, 1,   2,   2},
+	{DOLBY_CORE2A_CLKGATE_CTRL, 1,   4,   2},
+	{DOLBY_CORE3_CLKGATE_CTRL,  0,   1,   1},
+	{DOLBY_CORE3_CLKGATE_CTRL,  1,   2,   2},
+	{VPU_REG_END,               0,   0,   0},
+};
+
+static struct vpu_ctrl_s vpu_module_init_g12a[] = {
+	/* reg,                     val, bit, len */
+	{DOLBY_CORE1_CLKGATE_CTRL,  0x55555555, 0,   32},
+	{DOLBY_CORE2A_CLKGATE_CTRL, 0x55555555, 0,   32},
+	{DOLBY_CORE3_CLKGATE_CTRL,  0x55555555, 0,   32},
+	{VPU_REG_END,               0,          0,   0},
+};
+
+static struct vpu_ctrl_s vpu_module_init_tm2[] = {
+	/* reg,                     val, bit, len */
+	{DOLBY_CORE1_CLKGATE_CTRL,  0x55555555, 0,   32},
+	{DOLBY_CORE2A_CLKGATE_CTRL, 0x55555555, 0,   32},
+	{DOLBY_CORE3_CLKGATE_CTRL,  0x55555555, 0,   32},
+	{DOLBY_CORE1_1_CLKGATE_CTRL,0x55555555, 0,   32},
+	{DOLBY_TM2_TV_CLKGATE_CTRL, 0x55555555, 0,   32},
+	{VPU_REG_END,              0,          0,   0},
+};
+
+/* ******************************************************* */
+/*                 VPU reset table                    */
+/* ******************************************************* */
+static struct vpu_reset_s vpu_reset_gx[] = {
+	/* reg,        mask */
+	{RESET0_LEVEL, ((1<<5) | (1<<10) | (1<<19) | (1<<13))},
+	{RESET1_LEVEL, (1<<5)},
+	{RESET2_LEVEL, (1<<15)},
+	{RESET4_LEVEL, ((1<<6) | (1<<7) | (1<<13) | (1<<5) | (1<<9) | (1<<4) | (1<<12))},
+	{RESET7_LEVEL, (1<<7)},
+	{VPU_REG_END,  0},
+};
+
+static struct vpu_reset_s vpu_reset_txhd[] = {
+	/* reg,        mask */
+	{RESET0_LEVEL, ((1<<5) | (1<<10) | (1<<19) | (1<<13))},
+	{RESET1_LEVEL, (1<<5)},
+	{RESET2_LEVEL, (1<<15)},
+	{RESET4_LEVEL, ((1<<6) | (1<<7) | (1<<13) | (1<<5) | (1<<9) | (1<<4) | (1<<12))},
+	{RESET7_LEVEL, ((1<<7) | (1<<12))}, /* bit[12]:tcon */
+	{VPU_REG_END,  0},
+};
+
+static struct vpu_reset_s vpu_reset_tl1[] = {
+	/* reg,        mask */
+	{RESET0_LEVEL, ((1<<5) | (1<<10) | (1<<19) | (1<<13))},
+	{RESET1_LEVEL, ((1<<5) | (1<<4))},
+	{RESET2_LEVEL, (1<<15)},
+	{RESET4_LEVEL, ((1<<6) | (1<<7) | (1<<13) | (1<<5) | (1<<9) | (1<<4) | (1<<12))},
+	{RESET7_LEVEL, (1<<7)},
+	{VPU_REG_END,  0},
+};
+
+#endif
diff -ruN '--exclude=.git' '--exclude=build' '--exclude=.gitignore' '--exclude=.mailmap' '--exclude=*.o' '--exclude=*.a' '--exclude=*.pyc' '--exclude=aml-user-key.sig' '--exclude=aml-user-key-v3.sig' '--exclude=aml-key' a/drivers/vpu/aml_vpu_dummy_reg.h b/drivers/vpu/aml_vpu_dummy_reg.h
--- a/drivers/vpu/aml_vpu_dummy_reg.h	1970-01-01 08:00:00.000000000 +0800
+++ b/drivers/vpu/aml_vpu_dummy_reg.h	2026-02-09 11:25:48.633084759 +0800
@@ -0,0 +1,144 @@
+/* SPDX-License-Identifier: (GPL-2.0+ OR MIT) */
+/*
+ * drivers/vpu/aml_vpu_dummy_reg.h
+ *
+ * Copyright (C) 2020 Amlogic, Inc. All rights reserved.
+ *
+ */
+
+#ifndef __VPU_DUMMY_REG_H__
+#define __VPU_DUMMY_REG_H__
+#include <asm/arch/io.h>
+#include <asm/arch/secure_apb.h>
+
+#ifndef HHI_GP_PLL_CNTL
+#define HHI_GP_PLL_CNTL                            (0xc1100000 + (0x1010 << 2))
+#endif
+
+#ifndef HHI_GP1_PLL_CNTL
+#define HHI_GP1_PLL_CNTL                           (0xc883c000 + (0x16 << 2))
+#endif
+#ifndef HHI_GP1_PLL_CNTL2
+#define HHI_GP1_PLL_CNTL2                          (0xc883c000 + (0x17 << 2))
+#endif
+#ifndef HHI_GP1_PLL_CNTL3
+#define HHI_GP1_PLL_CNTL3                          (0xc883c000 + (0x18 << 2))
+#endif
+#ifndef HHI_GP1_PLL_CNTL4
+#define HHI_GP1_PLL_CNTL4                          (0xc883c000 + (0x19 << 2))
+#endif
+
+#ifndef HHI_VPU_MEM_PD_REG2
+#define HHI_VPU_MEM_PD_REG2                        (0xff63c000 + (0x4d << 2))
+#endif
+#ifndef HHI_VPU_MEM_PD_REG3
+#define HHI_VPU_MEM_PD_REG3                        (0xff63c000 + (0x4e << 2))
+#endif
+#ifndef HHI_VPU_MEM_PD_REG4
+#define HHI_VPU_MEM_PD_REG4                        (0xff63c000 + (0x4c << 2))
+#endif
+
+#ifndef DOLBY_TV_CLKGATE_CTRL
+#define DOLBY_TV_CLKGATE_CTRL                      (0x33f1)
+#endif
+#ifndef DOLBY_CORE1_CLKGATE_CTRL
+#define DOLBY_CORE1_CLKGATE_CTRL                   (0x33f2)
+#endif
+#ifndef DOLBY_CORE2A_CLKGATE_CTRL
+#define DOLBY_CORE2A_CLKGATE_CTRL                  (0x3432)
+#endif
+#ifndef DOLBY_CORE3_CLKGATE_CTRL
+#define DOLBY_CORE3_CLKGATE_CTRL                   (0x36f0)
+#endif
+#ifndef DOLBY_TM2_TV_CLKGATE_CTRL
+#define DOLBY_TM2_TV_CLKGATE_CTRL                  (0x43f1)
+#endif
+#ifndef DOLBY_CORE1_1_CLKGATE_CTRL
+#define DOLBY_CORE1_1_CLKGATE_CTRL                 (0x44f2)
+#endif
+
+#ifndef VENC_VDAC_TST_VAL
+#define VENC_VDAC_TST_VAL                          (0x1b7f)
+#endif
+#ifndef ENCP_DVI_HSO_BEGIN
+#define ENCP_DVI_HSO_BEGIN                         (0x1c30)
+#endif
+#ifndef VPP_DUMMY_DATA
+#define VPP_DUMMY_DATA                             (0x1d00)
+#endif
+#ifndef VPU_VPU_PWM_V0
+#define VPU_VPU_PWM_V0                             (0x2730)
+#endif
+
+#ifndef VPU_RDARB_MODE_L1C1
+#define VPU_RDARB_MODE_L1C1                        (0x2790)
+#endif
+#ifndef VPU_RDARB_MODE_L1C2
+#define VPU_RDARB_MODE_L1C2                        (0x2799)
+#endif
+
+#ifndef HHI_VPU_CLK_CNTL
+#define HHI_VPU_CLK_CNTL                           (0xc883c000 + (0x6f << 2))
+#endif
+
+#ifndef HHI_VPU_CLK_CTRL
+#define HHI_VPU_CLK_CTRL                           (0xc883c000 + (0x6f << 2))
+#endif
+
+#ifndef HHI_VAPBCLK_CNTL
+#define HHI_VAPBCLK_CNTL                           (0xc883c000 + (0x7d << 2))
+#endif
+
+#ifndef HHI_VID_CLK_CNTL2
+#define HHI_VID_CLK_CNTL2                          (0xff63c000 + (0x065 << 2))
+#endif
+
+#ifndef AO_RTI_GEN_PWR_SLEEP0
+#define AO_RTI_GEN_PWR_SLEEP0                      (0xff800000 + (0x03a << 2))
+#endif
+
+#ifndef AO_RTI_GEN_PWR_ISO0
+#define AO_RTI_GEN_PWR_ISO0                        (0xff800000 + (0x03b << 2))
+#endif
+
+#ifndef HHI_VPU_MEM_PD_REG0
+#define HHI_VPU_MEM_PD_REG0                        (0xff63c000 + (0x041 << 2))
+#endif
+
+#ifndef HHI_VPU_MEM_PD_REG1
+#define HHI_VPU_MEM_PD_REG1                        (0xff63c000 + (0x042 << 2))
+#endif
+
+#ifndef HHI_VPU_MEM_PD_REG2
+#define HHI_VPU_MEM_PD_REG2                        (0xff63c000 + (0x04d << 2))
+#endif
+
+#ifndef HHI_VPU_MEM_PD_REG3
+#define HHI_VPU_MEM_PD_REG3                        (0xff63c000 + (0x043 << 2))
+#endif
+
+#ifndef HHI_VPU_MEM_PD_REG4
+#define HHI_VPU_MEM_PD_REG4                        (0xff63c000 + (0x044 << 2))
+#endif
+
+#ifndef HHI_VPU_CLKB_CNTL
+#define HHI_VPU_CLKB_CNTL			   (0xff63c000 + (0x083 << 2))
+#endif
+
+#ifndef RESET0_LEVEL
+#define RESET0_LEVEL                               (0x0420)
+#endif
+#ifndef RESET1_LEVEL
+#define RESET1_LEVEL                               (0x0421)
+#endif
+#ifndef RESET2_LEVEL
+#define RESET2_LEVEL                               (0x0422)
+#endif
+#ifndef RESET4_LEVEL
+#define RESET4_LEVEL                               (0x0424)
+#endif
+#ifndef RESET7_LEVEL
+#define RESET7_LEVEL                               (0x0427)
+#endif
+
+#endif
diff -ruN '--exclude=.git' '--exclude=build' '--exclude=.gitignore' '--exclude=.mailmap' '--exclude=*.o' '--exclude=*.a' '--exclude=*.pyc' '--exclude=aml-user-key.sig' '--exclude=aml-user-key-v3.sig' '--exclude=aml-key' a/drivers/vpu/aml_vpu.h b/drivers/vpu/aml_vpu.h
--- a/drivers/vpu/aml_vpu.h	1970-01-01 08:00:00.000000000 +0800
+++ b/drivers/vpu/aml_vpu.h	2026-02-09 11:25:48.633084759 +0800
@@ -0,0 +1,131 @@
+/* SPDX-License-Identifier: (GPL-2.0+ OR MIT) */
+/*
+ * drivers/vpu/aml_vpu.h
+ *
+ * Copyright (C) 2020 Amlogic, Inc. All rights reserved.
+ *
+ */
+
+#ifndef __VPU_PARA_H__
+#define __VPU_PARA_H__
+#ifdef CONFIG_SECURE_POWER_CONTROL
+#include <asm/arch/pwr_ctrl.h>
+#endif
+
+//#define VPU_DEBUG_PRINT
+#define VPUPR(fmt, args...)     printf("vpu: "fmt"", ## args)
+#define VPUERR(fmt, args...)    printf("vpu: error: "fmt"", ## args)
+
+enum vpu_chip_e {
+	VPU_CHIP_GXBB = 0,
+	VPU_CHIP_GXTVBB, /* 1 */
+	VPU_CHIP_GXL, /* 2 */
+	VPU_CHIP_GXM, /* 3 */
+	VPU_CHIP_TXL, /* 4 */
+	VPU_CHIP_TXLX, /* 5 */
+	VPU_CHIP_AXG, /* 6 */
+	VPU_CHIP_TXHD, /* 7 */
+	VPU_CHIP_G12A, /* 8 */
+	VPU_CHIP_G12B, /* 9 */
+	VPU_CHIP_TL1, /* 10 */
+	VPU_CHIP_SM1, /* 11 */
+	VPU_CHIP_TM2, /* 12 */
+	VPU_CHIP_T5, /* 13 */
+	VPU_CHIP_T5D, /* 14 */
+	VPU_CHIP_T5W, /* 15 */
+	VPU_CHIP_MAX,
+};
+
+#define VPU_PWR_ON             1
+#define VPU_PWR_OFF            0
+#define VPU_PWR_ID_INVALID     0xffff
+#ifndef PM_VPU_HDMI
+#define PM_VPU_HDMI            5
+#endif
+
+//extern enum vpu_chip_e vpu_chip_type;
+
+#define VPU_REG_END            0xffff
+#define VPU_MEM_PD_CNT_MAX     10
+#define VPU_ISO_CNT_MAX        5
+#define VPU_PWR_CNT_MAX        5
+#define VPU_RESET_CNT_MAX      10
+
+struct fclk_div_s {
+	unsigned int fclk_id;
+	unsigned int fclk_mux;
+	unsigned int fclk_div;
+};
+
+struct vpu_clk_s {
+	unsigned int freq;
+	unsigned int mux;
+	unsigned int div;
+};
+
+struct vpu_ctrl_s {
+	unsigned int reg;
+	unsigned int val;
+	unsigned int bit;
+	unsigned int len;
+};
+
+struct vpu_reset_s {
+	unsigned int reg;
+	unsigned int mask;
+};
+
+struct vpu_data_s {
+	enum vpu_chip_e chip_type;
+	const char *chip_name;
+	unsigned char clk_level_dft;
+	unsigned char clk_level_max;
+	unsigned char gp_pll_valid;
+
+	unsigned int vpu_clk_reg;
+	unsigned int vapb_clk_reg;
+	unsigned int vid_clk_reg;
+
+	unsigned int pwrctrl_id;
+
+	struct fclk_div_s *fclk_div_table;
+	struct vpu_clk_s  *vpu_clk_table;
+
+	struct vpu_ctrl_s *mem_pd_table;
+	struct vpu_ctrl_s *power_table;
+	struct vpu_ctrl_s *iso_table;
+	struct vpu_reset_s *reset_table;
+
+	unsigned int module_init_table_cnt;
+	struct vpu_ctrl_s *module_init_table;
+
+	void (*power_on)(void);
+	void (*power_off)(void);
+};
+
+struct vpu_conf_s {
+	struct vpu_data_s *data;
+	char drv_version[20];
+
+	unsigned int clk_level;
+	unsigned int fclk_freq; /* unit: MHz */
+};
+
+extern struct vpu_conf_s vpu_conf;
+
+/* ************************************************ */
+/* extern function */
+/* ************************************************ */
+extern void udelay(unsigned long usec);
+extern int printf(const char *fmt, ...);
+/* ************************************************ */
+
+extern void vpu_mem_pd_init_off(void);
+extern void vpu_module_init_config(void);
+
+extern void vpu_power_on(void);
+extern void vpu_power_off(void);
+void vpu_power_on_new(void);
+void vpu_power_off_new(void);
+
+#endif
diff -ruN '--exclude=.git' '--exclude=build' '--exclude=.gitignore' '--exclude=.mailmap' '--exclude=*.o' '--exclude=*.a' '--exclude=*.pyc' '--exclude=aml-user-key.sig' '--exclude=aml-user-key-v3.sig' '--exclude=aml-key' a/drivers/vpu/aml_vpu_power_init.c b/drivers/vpu/aml_vpu_power_init.c
--- a/drivers/vpu/aml_vpu_power_init.c	1970-01-01 08:00:00.000000000 +0800
+++ b/drivers/vpu/aml_vpu_power_init.c	2026-02-09 11:25:48.633084759 +0800
@@ -0,0 +1,259 @@
+/* SPDX-License-Identifier: (GPL-2.0+ OR MIT) */
+/*
+ * drivers/vpu/aml_vpu_power_init.c
+ *
+ * Copyright (C) 2020 Amlogic, Inc. All rights reserved.
+ *
+ */
+
+#include <config.h>
+#include <linux/kernel.h>
+#include <vpu.h>
+#include "aml_vpu_reg.h"
+#include "aml_vpu.h"
+
+void vpu_mem_pd_init_off(void)
+{
+	return;
+#ifdef VPU_DEBUG_PRINT
+	VPUPR("%s\n", __func__);
+#endif
+}
+
+void vpu_module_init_config(void)
+{
+	struct vpu_ctrl_s *ctrl_table;
+	unsigned int _reg, _val, _bit, _len;
+	int i = 0, cnt;
+
+	/* vpu clk gate init off */
+	cnt = vpu_conf.data->module_init_table_cnt;
+	ctrl_table = vpu_conf.data->module_init_table;
+	if (ctrl_table) {
+		i = 0;
+		while (i < cnt) {
+			if (ctrl_table[i].reg == VPU_REG_END)
+				break;
+			_reg = ctrl_table[i].reg;
+			_val = ctrl_table[i].val;
+			_bit = ctrl_table[i].bit;
+			_len = ctrl_table[i].len;
+			vpu_vcbus_setb(_reg, _val, _bit, _len);
+			i++;
+		}
+	}
+	vpu_hiu_setb(vpu_conf.data->vid_clk_reg, 0, 0, 8);
+
+	/* dmc_arb_config */
+	switch (vpu_conf.data->chip_type) {
+	case VPU_CHIP_GXBB:
+	case VPU_CHIP_GXTVBB:
+	case VPU_CHIP_GXL:
+	case VPU_CHIP_GXM:
+	case VPU_CHIP_TXL:
+	case VPU_CHIP_TXLX:
+	case VPU_CHIP_AXG:
+	case VPU_CHIP_TXHD:
+	case VPU_CHIP_G12A:
+	case VPU_CHIP_G12B:
+	case VPU_CHIP_SM1:
+		vpu_vcbus_write(VPU_RDARB_MODE_L1C1, 0x0); //0x210000
+		vpu_vcbus_write(VPU_RDARB_MODE_L1C2, 0x10000);
+		vpu_vcbus_write(VPU_RDARB_MODE_L2C1, 0x900000);
+		vpu_vcbus_write(VPU_WRARB_MODE_L2C1, 0x20000);
+		break;
+	case VPU_CHIP_TL1:
+	case VPU_CHIP_TM2:
+	case VPU_CHIP_T5:
+	case VPU_CHIP_T5D:
+		vpu_vcbus_write(VPU_RDARB_MODE_L1C1, 0x0); //0x210000
+		vpu_vcbus_write(VPU_RDARB_MODE_L1C2, 0x10000);
+		vpu_vcbus_write(VPU_RDARB_MODE_L2C1, 0x20000);
+		vpu_vcbus_write(VPU_WRARB_MODE_L2C1, 0x20000);
+		break;
+	default:
+		break;
+	}
+
+	VPUPR("%s\n", __func__);
+}
+
+void vpu_power_on(void)
+{
+	struct vpu_ctrl_s *ctrl_table;
+	struct vpu_reset_s *reset_table;
+	unsigned int _reg, _val, _start, _end, _len, mask;
+	int i = 0, j;
+
+	/* power on VPU_HDMI */
+	ctrl_table = vpu_conf.data->power_table;
+	if (ctrl_table) {
+		i = 0;
+		while (i < VPU_PWR_CNT_MAX) {
+			if (ctrl_table[i].reg == VPU_REG_END)
+				break;
+			_reg = ctrl_table[i].reg;
+			_val = 0;
+			_start = ctrl_table[i].bit;
+			_len = ctrl_table[i].len;
+			vpu_ao_setb(_reg, _val, _start, _len);
+			i++;
+		}
+	}
+	udelay(20);
+
+	/* power up memories */
+	ctrl_table = vpu_conf.data->mem_pd_table;
+	i = 0;
+	while (i < VPU_MEM_PD_CNT_MAX) {
+		if (ctrl_table[i].reg == VPU_REG_END)
+			break;
+		_reg = ctrl_table[i].reg;
+		_start = ctrl_table[i].bit;
+		_end = ctrl_table[i].len + ctrl_table[i].bit;
+		for (j = _start; j < _end; j+=2) {
+			vpu_hiu_setb(_reg, 0, j, 2);
+			udelay(5);
+		}
+		i++;
+	}
+	for (i = 8; i < 16; i++) {
+		vpu_hiu_setb(HHI_MEM_PD_REG0, 0, i, 1);
+		udelay(5);
+	}
+	udelay(20);
+
+	/* Reset VIU + VENC */
+	/* Reset VENCI + VENCP + VADC + VENCL */
+	/* Reset HDMI-APB + HDMI-SYS + HDMI-TX + HDMI-CEC */
+	reset_table = vpu_conf.data->reset_table;
+	i = 0;
+	while (i < VPU_RESET_CNT_MAX) {
+		if (reset_table[i].reg == VPU_REG_END)
+			break;
+		_reg = reset_table[i].reg;
+		mask = reset_table[i].mask;
+		vpu_cbus_clr_mask(_reg, mask);
+		i++;
+	}
+	udelay(5);
+	/* release Reset */
+	i = 0;
+	while (i < VPU_RESET_CNT_MAX) {
+		if (reset_table[i].reg == VPU_REG_END)
+			break;
+		_reg = reset_table[i].reg;
+		mask = reset_table[i].mask;
+		vpu_cbus_set_mask(_reg, mask);
+		i++;
+	}
+
+	/* Remove VPU_HDMI ISO */
+	ctrl_table = vpu_conf.data->iso_table;
+	if (ctrl_table) {
+		i = 0;
+		while (i < VPU_ISO_CNT_MAX) {
+			if (ctrl_table[i].reg == VPU_REG_END)
+				break;
+			_reg = ctrl_table[i].reg;
+			_val = 0;
+			_start = ctrl_table[i].bit;
+			_len = ctrl_table[i].len;
+			vpu_ao_setb(_reg, _val, _start, _len);
+			i++;
+		}
+	}
+
+	VPUPR("%s\n", __func__);
+}
+
+void vpu_power_off(void)
+{
+	struct vpu_ctrl_s *ctrl_table;
+	unsigned int _reg, _start, _end, _len, _val;
+	int i = 0, j;
+
+	/* Enable Isolation */
+	ctrl_table = vpu_conf.data->iso_table;
+	if (ctrl_table) {
+		i = 0;
+		while (i < VPU_ISO_CNT_MAX) {
+			if (ctrl_table[i].reg == VPU_REG_END)
+				break;
+			_reg = ctrl_table[i].reg;
+			_val = 1;
+			_start = ctrl_table[i].bit;
+			_len = ctrl_table[i].len;
+			vpu_ao_setb(_reg, _val, _start, _len);
+			i++;
+		}
+	}
+	udelay(20);
+
+	/* power down memories */
+	ctrl_table = vpu_conf.data->mem_pd_table;
+	i = 0;
+	while (i < VPU_MEM_PD_CNT_MAX) {
+		if (ctrl_table[i].reg == VPU_REG_END)
+			break;
+		_reg = ctrl_table[i].reg;
+		_start = ctrl_table[i].bit;
+		_end = ctrl_table[i].len + ctrl_table[i].bit;
+		for (j = _start; j < _end; j+=2) {
+			vpu_hiu_setb(_reg, 0x3, j, 2);
+			udelay(5);
+		}
+		i++;
+	}
+	for (i = 8; i < 16; i++) {
+		vpu_hiu_setb(HHI_MEM_PD_REG0, 0x1, i, 1);
+		udelay(5);
+	}
+	udelay(20);
+
+	/* Power down VPU domain */
+	ctrl_table = vpu_conf.data->power_table;
+	if (ctrl_table) {
+		i = 0;
+		while (i < VPU_PWR_CNT_MAX) {
+			if (ctrl_table[i].reg == VPU_REG_END)
+				break;
+			_reg = ctrl_table[i].reg;
+			_val = ctrl_table[i].val;
+			_start = ctrl_table[i].bit;
+			_len = ctrl_table[i].len;
+			vpu_ao_setb(_reg, _val, _start, _len);
+			i++;
+		}
+	}
+
+	vpu_hiu_setb(vpu_conf.data->vapb_clk_reg, 0, 8, 1);
+	vpu_hiu_setb(vpu_conf.data->vpu_clk_reg, 0, 8, 1);
+
+	VPUPR("%s\n", __func__);
+}
+
+void vpu_power_on_new(void)
+{
+#ifdef CONFIG_SECURE_POWER_CONTROL
+	if (vpu_conf.data->pwrctrl_id < VPU_PWR_ID_INVALID)
+		pwr_ctrl_psci_smc(vpu_conf.data->pwrctrl_id, 1);
+	VPUPR("%s\n", __func__);
+#else
+	VPUERR("%s: no CONFIG_SECURE_POWER_CONTROL\n", __func__);
+#endif
+}
+
+void vpu_power_off_new(void)
+{
+#ifdef CONFIG_SECURE_POWER_CONTROL
+	VPUPR("%s\n", __func__);
+	if (vpu_conf.data->pwrctrl_id < VPU_PWR_ID_INVALID)
+		pwr_ctrl_psci_smc(vpu_conf.data->pwrctrl_id, 0);
+#else
+	VPUERR("%s: no CONFIG_SECURE_POWER_CONTROL\n", __func__);
+#endif
+
+	vpu_hiu_setb(vpu_conf.data->vapb_clk_reg, 0, 8, 1);
+	vpu_hiu_setb(vpu_conf.data->vpu_clk_reg, 0, 8, 1);
+}
diff -ruN '--exclude=.git' '--exclude=build' '--exclude=.gitignore' '--exclude=.mailmap' '--exclude=*.o' '--exclude=*.a' '--exclude=*.pyc' '--exclude=aml-user-key.sig' '--exclude=aml-user-key-v3.sig' '--exclude=aml-key' a/drivers/vpu/aml_vpu_reg.h b/drivers/vpu/aml_vpu_reg.h
--- a/drivers/vpu/aml_vpu_reg.h	1970-01-01 08:00:00.000000000 +0800
+++ b/drivers/vpu/aml_vpu_reg.h	2026-02-09 11:25:48.633084759 +0800
@@ -0,0 +1,150 @@
+/* SPDX-License-Identifier: (GPL-2.0+ OR MIT) */
+/*
+ * drivers/vpu/aml_vpu_reg.h
+ *
+ * Copyright (C) 2020 Amlogic, Inc. All rights reserved.
+ *
+ */
+
+#ifndef __VPU_REG_H__
+#define __VPU_REG_H__
+#include <asm/arch/io.h>
+#include <asm/arch/secure_apb.h>
+#ifdef CONFIG_AML_VPU_REG_NEW
+#include <asm/arch/vpu_regs.h>
+#endif
+#include "aml_vpu.h"
+#include "aml_vpu_dummy_reg.h"
+
+/* ********************************
+ * register define
+ * ********************************* */
+/* offset */
+#define REG_OFFSET_CBUS(reg)        (((reg) << 2))
+#define REG_OFFSET_VCBUS(reg)       (((reg) << 2))
+
+/* memory mapping */
+#define REG_ADDR_AOBUS(reg)         (reg + 0L)
+#define REG_ADDR_HIU(reg)           (reg + 0L)
+#define REG_ADDR_CBUS(reg)          (REG_BASE_CBUS + REG_OFFSET_CBUS(reg))
+#define REG_ADDR_VCBUS(reg)         (REG_BASE_VCBUS + REG_OFFSET_VCBUS(reg))
+
+/* ********************************
+ * register access api
+ * ********************************* */
+static inline unsigned int vpu_hiu_read(unsigned int _reg)
+{
+	unsigned int val = 0;
+
+	val = *(volatile unsigned int *)(REG_ADDR_HIU(_reg));
+
+	return val;
+};
+
+static inline void vpu_hiu_write(unsigned int _reg, unsigned int _value)
+{
+	*(volatile unsigned int *)REG_ADDR_HIU(_reg) = (_value);
+};
+
+static inline void vpu_hiu_setb(unsigned int _reg, unsigned int _value,
+		unsigned int _start, unsigned int _len)
+{
+	vpu_hiu_write(_reg, ((vpu_hiu_read(_reg) &
+			~(((1L << (_len))-1) << (_start))) |
+			(((_value)&((1L<<(_len))-1)) << (_start))));
+}
+
+static inline unsigned int vpu_hiu_getb(unsigned int _reg,
+		unsigned int _start, unsigned int _len)
+{
+	return (vpu_hiu_read(_reg) >> (_start)) & ((1L << (_len)) - 1);
+}
+
+static inline void vpu_hiu_set_mask(unsigned int _reg, unsigned int _mask)
+{
+	vpu_hiu_write(_reg, (vpu_hiu_read(_reg) | (_mask)));
+}
+
+static inline void vpu_hiu_clr_mask(unsigned int _reg, unsigned int _mask)
+{
+	vpu_hiu_write(_reg, (vpu_hiu_read(_reg) & (~(_mask))));
+}
+
+static inline unsigned int vpu_vcbus_read(unsigned int _reg)
+{
+	unsigned int val = 0;
+	val = *(volatile unsigned int *)(REG_ADDR_VCBUS(_reg));
+	return val;
+};
+
+static inline void vpu_vcbus_write(unsigned int _reg, unsigned int _value)
+{
+	*(volatile unsigned int *)REG_ADDR_VCBUS(_reg) = (_value);
+};
+
+static inline void vpu_vcbus_setb(unsigned int _reg, unsigned int _value,
+		unsigned int _start, unsigned int _len)
+{
+	vpu_vcbus_write(_reg, ((vpu_vcbus_read(_reg) &
+			~(((1L << (_len))-1) << (_start))) |
+			(((_value)&((1L<<(_len))-1)) << (_start))));
+}
+
+static inline unsigned int vpu_cbus_read(unsigned int _reg)
+{
+	unsigned int val = 0;
+	val = *(volatile unsigned int *)(REG_ADDR_CBUS(_reg));
+	return val;
+};
+
+static inline void vpu_cbus_write(unsigned int _reg, unsigned int _value)
+{
+	*(volatile unsigned int *)REG_ADDR_CBUS(_reg) = (_value);
+};
+
+static inline void vpu_cbus_setb(unsigned int _reg, unsigned int _value,
+		unsigned int _start, unsigned int _len)
+{
+	vpu_cbus_write(_reg, ((vpu_cbus_read(_reg) &
+			~(((1L << (_len))-1) << (_start))) |
+			(((_value)&((1L<<(_len))-1)) << (_start))));
+}
+
+static inline unsigned int vpu_cbus_getb(unsigned int _reg,
+		unsigned int _start, unsigned int _len)
+{
+	return (vpu_cbus_read(_reg) >> (_start)) & ((1L << (_len)) - 1);
+}
+
+static inline void vpu_cbus_set_mask(unsigned int _reg, unsigned int _mask)
+{
+	vpu_cbus_write(_reg, (vpu_cbus_read(_reg) | (_mask)));
+}
+
+static inline void vpu_cbus_clr_mask(unsigned int _reg, unsigned int _mask)
+{
+	vpu_cbus_write(_reg, (vpu_cbus_read(_reg) & (~(_mask))));
+}
+
+static inline unsigned int vpu_aobus_read(unsigned int _reg)
+{
+	unsigned int val = 0;
+
+	val = *(volatile unsigned int *)(REG_ADDR_AOBUS(_reg));
+	return val;
+};
+
+static inline void vpu_aobus_write(unsigned int _reg, unsigned int _value)
+{
+	*(volatile unsigned int *)REG_ADDR_AOBUS(_reg) = (_value);
+};
+
+static inline void vpu_ao_setb(unsigned int _reg, unsigned int _value,
+		unsigned int _start, unsigned int _len)
+{
+	vpu_aobus_write(_reg, ((vpu_aobus_read(_reg) &
+			~(((1L << (_len))-1) << (_start))) |
+			(((_value)&((1L<<(_len))-1)) << (_start))));
+}
+
+#endif
diff -ruN '--exclude=.git' '--exclude=build' '--exclude=.gitignore' '--exclude=.mailmap' '--exclude=*.o' '--exclude=*.a' '--exclude=*.pyc' '--exclude=aml-user-key.sig' '--exclude=aml-user-key-v3.sig' '--exclude=aml-key' a/drivers/vpu/Makefile b/drivers/vpu/Makefile
--- a/drivers/vpu/Makefile	1970-01-01 08:00:00.000000000 +0800
+++ b/drivers/vpu/Makefile	2026-02-09 11:25:48.633084759 +0800
@@ -0,0 +1,2 @@
+obj-$(CONFIG_AML_VPU) += aml_vpu.o aml_vpu_power_init.o
+
